node/node.go

```diff
@@ -264,30 +274,68 @@ func New(host host.Host, consensus *bft.Consensus, db *ethdb.LDBDatabase) *Node
 	// Setup initial state of syncing.
 	node.syncingState = NotDoingSyncing
 	node.StopPing = make(chan struct{})
+	node.Consensus.ConsensusBlock = make(chan *types.Block)
+	node.Consensus.VerifiedNewBlock = make(chan *types.Block)
+	node.peerRegistrationRecord = make(map[uint32]*syncConfig)
 
 	node.OfflinePeers = make(chan p2p.Peer)
 	go node.RemovePeersHandler()
 
 	return &node
 }
 
-// DoSyncing starts syncing.
-func (node *Node) DoSyncing() {
-	// If this node is currently doing sync, another call for syncing will be returned immediately.
-	if !atomic.CompareAndSwapUint32(&node.syncingState, NotDoingSyncing, DoingSyncing) {
-		return
+// IsOutOfSync checks whether the node is out of sync by comparing latest block with consensus block
+func (node *Node) IsOutOfSync(consensusBlock *types.Block) bool {
+	myHeight := node.blockchain.CurrentBlock().NumberU64()
+	newHeight := consensusBlock.NumberU64()
+	node.log.Debug("[sync]", "myHeight", myHeight, "newHeight", newHeight)
+	if newHeight-myHeight <= 2 {
+		return false
 	}
-	defer atomic.StoreUint32(&node.syncingState, NotDoingSyncing)
-	if node.stateSync == nil {
-		node.stateSync = syncing.GetStateSync()
+	// cache latest blocks for last mile catch up
+	if newHeight-myHeight <= 4 && node.stateSync != nil {
```

Comment:
Constant again. Why 4? -

There are many details of syncing about constants 2 and 4. Can you write a short syncing.md about the state machine and the constants

Reply:
I changed it to variable. I will write some comments in syncing.md